crossplane/crossplane
# Introduce Crossplane version constraints in packages

### Description of your changes



Adds Crossplane version constraints to packages with the following behavior:
- Packages that have constraints that do not match the current Crossplane version will not be installed successfully (i.e. no resources will be created)
- Crank will check the `crossplane` spec field during build to ensure the semver constrains provided are valid ranges
- Specifying `ignoreCrossplaneConstraints: true` on `Provider` or `Configuration` spec will cause package revision controllers to skip checking crossplane constrains (useful for dev when using `go run` and version is empty)
- Revision controllers will refuse to install a package with invalid version constraints, even when `ignoreCrossplaneConstraints: true` (note the package would have had to be built with tool other than crank given that crank will validate version constraints)

For documentation on semver constraint format, see https://github.com/Masterminds/semver

Fixes #1843 

I have:


### How has this code been tested



1. `make build`
2. Install locally built Crossplane helm chart
3. Install package with following `crossplane.yaml`

```yaml
apiVersion: meta.pkg.crossplane.io/v1alpha1
kind: Provider
metadata:
  name: test-configuration
spec:
  crossplane: ">=1.0.0"
````

4. Observe failure to install and the following event:

```yaml
  Type     Reason       Age              From                                              Message
  ----     ------       ----             ----                                              -------
  Warning  LintPackage  8s (x3 over 9s)  packages/configurationrevision.pkg.crossplane.io  package is not compatible with Crossplane version (v0.14.0-rc.26.gfa9e89fd)
```

5. Delete package revision and install with `ignoreCrossplaneConstraints: true`

```yaml
apiVersion: pkg.crossplane.io/v1alpha1
kind: Configuration
metadata:
  name: test-configuration
spec:
  package: hasheddan/test-configuration:latest
  ignoreCrossplaneConstraints: true
```

6. Observe successful install

```yaml
NAME                                                 INSTALLED   HEALTHY   PACKAGE                               AGE
configuration.pkg.crossplane.io/test-configuration   True        True      hasheddan/test-configuration:latest   11s
```

7. Also installed some providers, which currently don't have any Crossplane version constraints and they installed successfully.

---


Crank behavior:

1. Build package with `crossplane.yaml` and locally built crank

```yaml
apiVersion: meta.pkg.crossplane.io/v1alpha1
kind: Configuration
metadata:
  name: test-configuration
spec:
  crossplane: ">=a1.0.0"
```

2. Observe following error:

```
kubectl crossplane: error: failed to build package: failed to lint package: package version constraints are poorly formatted: improper constraint: >=a1.0.0
```



[contribution process]: https://git.io/fj2m9
